src/gluonts/torch/distributions/mqf2.py

```diff
@@ -0,0 +1,395 @@
+        where ES(g,z_i) =
+        -1/(2*es_num_samples^2) * sum_{w,w'} ||w-w'||_2^beta + 1/es_num_samples * sum_{w''} ||w''-z_i||_2^beta,
+        w's are samples drawn from the quantile function g(., h_i) (gradient of picnn),
+        h_i is the hidden state associated with z_i,
+        and es_num_samples is the number of samples drawn for each of w, w', w'' in energy score approximation
+
+        Parameters
+        ----------
+        z
+            A batch of time series with shape (batch_size, context_length + prediction_length - 1)
+
+        Returns
+        -------
+        loss
+            Tensor of shape (batch_size * context_length,)
+        """
+
+        es_num_samples = self.es_num_samples
+        hidden_state = self.hidden_state
+        picnn = self.picnn
+        beta = self.beta
+
+        z = self.unfold_observations(z)
+        hidden_state = hidden_state.reshape(-1, hidden_state.shape[-1])
+
+        loss = picnn.energy_score(
+            z, hidden_state, es_num_samples=es_num_samples, beta=beta
+        )
+
+        return loss
+
+    def rsample(self, sample_shape: torch.Size = torch.Size()) -> torch.Tensor:
+        """
+        Generates the sample paths
+
+        Parameters
+        ----------
+        sample_shape
+            Shape of the samples
+
+        Returns
+        -------
+        sample_paths
+            Tesnor of shape (batch_size, *sample_shape, prediction_length)
+        """
+
+        numel_batch = self.numel_batch
+        hidden_state = self.hidden_state
+        prediction_length = self.prediction_length
+
+        num_samples_per_batch = MQF2Distribution.get_numel(sample_shape)
+        num_samples = num_samples_per_batch * numel_batch
+
+        hidden_state_repeat = hidden_state.repeat_interleave(
+            repeats=num_samples_per_batch, dim=0
+        )
+
+        alpha = torch.rand(
+            (num_samples, prediction_length),
+            dtype=hidden_state.dtype,
+            device=hidden_state.device,
+            layout=hidden_state.layout,
+        )
+
+        return self.quantile(alpha, hidden_state_repeat).reshape(
+            (numel_batch,) + sample_shape + (prediction_length,)
+        )
+
+    def quantile(
+        self, alpha: torch.Tensor, hidden_state: Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        """
+        Generates the predicted paths associated with the quantile levels alpha
+
+        Parameters
+        ----------
+        alpha
+            quantile levels,
+            shape = (batch_shape, prediction_length)
+        hidden_state
+            hidden_state, shape = (batch_shape, hidden_size)
+
+        Returns
+        -------
+        results
+            predicted paths of shape = (batch_shape, prediction_length)
+        """
+
+        picnn = self.picnn
+        if hidden_state is None:
+            hidden_state = self.hidden_state
+
+        normal_quantile = self.standard_normal.icdf(alpha)
+
+        if self.is_energy_score:
```

Comment:
I think some explanation would be useful here.

Reply:
Yes I will add an explanation